### PR TITLE
ZCS-6695 Fixing deserialization to only whitelisted classes

### PR DESCRIPTION
**Pull request by rupalid**
Fixing ObjectInputStream to deserialize only whitelisted classes
Followed the techniques specified in https://www.owasp.org/index.php/Deserialization_Cheat_Sheet#Java
Add a SecureObjectInputStream to only deserialize white listed classes
Add an LDAP config which holds a list of classes deserializer methods currently deserialize

Testing:
Added unit test to verify that an incorrect object during deserialization causes exception and is not read
Installed the code changes on ZCS and verified that current functionality is not affected
Verified that IMAP functionality is not affected
Verified no new exceptions are thrown in the logs

Tests | Failures | Errors | Skipped | Success rate | Time
1319 | 0 | 0 | 10 | 100.00% | 2567.717

To be done:
Complete SOAP automation

### CODE REVIEW

File: store/src/java/com/zimbra/cs/io/SecureObjectInputStream.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        try {
+             config = Provisioning.getInstance().getConfig().getDeserializerWhiteList();
+            for (int i = 0; i < config.length; ++i) {
+                if (desc.getName().equals(config[i].getClass().getName())) {
+                    return super.resolveClass(desc);
+                }
+            }
+            throw new InvalidClassException( "Unauthorized deserialization attempt", desc.getName());
+        } catch (ServiceException e) {
+            ZimbraLog.misc.errorQuietly("Error fetching  white list of deseializable  classes", e);
```
Comment (plobbes) [inline store/src/java/com/zimbra/cs/io/SecureObjectInputStream.java]: nit: extra spaces and typo in error message

**Comment by rupalid** [comment]
> I like the idea, although I'm hesitant to have this be configurable. This isn't a run time decision we're making, it's a predetermined set of things we allow, so why make it configurable? I vote we make the only allowable type valid in the context that we expect to see them and not have a generic (configurable) set of possible object types. Is that overly simplistic and/or paranoid on my part?

Fixed.

File: store/src/java/com/zimbra/cs/account/ZAttrConfig.java [hunk 2]
```diff
@@ -17973,7 +17973,7 @@ public void unsetGalLdapAttrMap() throws com.zimbra.common.service.ServiceExcept
...
..."adAutoComplete:(&(|(displayName=%s*)(cn=%s*)(sn=%s*)(givenName=%s*)(mail=%s*))(!(msExchHideFromAddressLists=TRUE))(|(&(objectCategory=person)(objectClass=user)(!(homeMDB=*))(!(msExchHomeServerName=*)))(&(objectCategory=person)(objectClass=user)(|(homeMDB=*)(msExchHomeServerName=*)))(&(objectCategory=person)(objectClass=contact))(objectCategory=group)(objectCategory=publicFolder)(objectCategory=msExchDynamicDistributionList)))","externalLdapAutoComplete:(|(cn=%s*)(sn=%s*)(gn=%s*)(mail=%s*))","email_has:(mail=*%s*)","email2_has:(mail=*%s*)","email3_has:(mail=*%s*)","email4_has:(mail=*%s*)","email5_has:(mail=*%s*)","email6_has:(mail=*%s*)","email7_has:(mail=*%s*)","email8_has:(mail=*%s*)","email9_has:(mail=*%s*)","email10_has:(mail=*%s*)","email11_has:(mail=*%s*)","email12_has:(mail=*%s*)","email13_has:(mail=*%s*)","email14_has:(mail=*%s*)","email15_has:(mail=*%s*)","email16_has:(mail=*%s*)","department_has:(ou=*%s*)","firstName_has:(gn=*%s*)","lastName_has:(sn=*%s*)","middleName_has:(initials=*%s*)","nickname_has:(|(displayName=*%s*)(cn=*%s*))","phoneticFirstName_has:(zimbraPhoneticFirstName=*%s*)","phoneticLastName_has:(zimbraPhoneticLastName=*%s*)","notes_has:(description=*%s*)"};
```
Comment (plobbes) [inline store/src/java/com/zimbra/cs/account/ZAttrConfig.java]: Hmm.  Does this change belong with this fix?  Seems unrelated.
